Loop/Managers/CriticalEventLogExportManager.swift

```diff
@@ -0,0 +1,513 @@
+
+    /// Export
+    /// - Parameter now: - The current time.
+    /// - Returns: Any error that occurred during export, including cancelled.
+    func export(now: Date) -> Error?
+}
+
+extension CriticalEventLogExporter {
+
+    /// Export
+    /// - Returns: Any error that occurred during export, including cancelled.
+    func export() -> Error? {
+        return export(now: Date())
+    }
+}
+
+public protocol CriticalEventLogExporterDelegate: AnyObject {
+
+    /// Some progress was made towards the export.
+    /// - Parameter progress: The current percent progress made (0.0 to 1.0) to complete the export.
+    func exportDidProgress(_ progress: Double)
+}
+
+// MARK: - CriticalEventLogExportManager
+
+fileprivate protocol CriticalEventLogSynchronizedExporter: CriticalEventLogExporter {
+    func exportSynchronized(now: Date) -> Error?
+}
+
+public class CriticalEventLogExportManager {
+    public let logs: [CriticalEventLog]
+    public let directory: URL
+    public let historicalDuration: TimeInterval
+    public let fileManager: FileManager
+
+    private var synchronizeSemaphore = DispatchSemaphore(value: 1)
+    private let lockedActiveExporter = Locked<CriticalEventLogSynchronizedExporter?>(nil)
+
+    private let log = OSLog(category: "CriticalEventLogExportManager")
+
+    public init(logs: [CriticalEventLog], directory: URL, historicalDuration: TimeInterval, fileManager: FileManager = FileManager.default) {
+        self.logs = logs
+        self.directory = directory
+        self.historicalDuration = historicalDuration
+        self.fileManager = fileManager
+    }
+
+    public func nextExportHistoricalDate(now: Date = Date()) -> Date {
+        switch latestExportDate() {
+        case .failure(let error):
+            log.error("Failure determining next export historical date: %{public}@", String(describing: error))
+        case .success(let latestExportDate):
+            if latestExportDate >= recentDate(from: now) {
+                return exportDate(for: date(byAddingDays: 1, to: now))
+            }
+        }
+        return now
+    }
+
+    private let retryExportHistoricalDuration: TimeInterval = .hours(1)
+
+    public func retryExportHistoricalDate(now: Date = Date()) -> Date {
+        return now.addingTimeInterval(retryExportHistoricalDuration)
+    }
+
+    // MARK: - Exporter
+
+    public func createExporter(to url: URL) -> CriticalEventLogExporter {
+        return CriticalEventLogFullExporter(manager: self, to: url)
+    }
+
+    public func createHistoricalExporter() -> CriticalEventLogExporter {
+        return CriticalEventLogHistoricalExporter(manager: self)
+    }
+
+    // MARK: - Export synchronization
+
+    private let synchronizeExportExpireDuration: TimeInterval = .seconds(5)
+    private let synchronizeExportCancelDuration: TimeInterval = .seconds(0.1)
+
+    fileprivate func synchronizeExport(for exporter: CriticalEventLogSynchronizedExporter, cancellingActive: Bool, now: Date) -> Error? {
```

Comment:
I'm not exactly sure why this function uses a semaphore.  Why is this needed?  I could use some help understanding it.

Reply:
The semaphore is necessary to coordinate the two exports that could be operating at once. The historical export could be running in the background when the user requests a full export. The semaphore allows the wait for the historical export to cancel before the full export begins. (Running more than one export at a time would be a waste of resources, if not worse.) The semaphore enables this coordination.

Reply:
✅